A2017/tensorflow.ipynb

```diff
@@ -0,0 +1,892 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1 + 3 = 4\n"
+     ]
+    }
+   ],
+   "source": [
+    "one_plus_three = tf.add(1, 3)\n",
+    "\n",
+    "with tf.Session() as session:\n",
+    "    print('1 + 3 = {}'.format(session.run(one_plus_three)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Gradients"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "TensorFlow supports _automatic differentiation_, i.e., it can compute the derivative of scalars with respect to tensors in the graph and represent the result as a symbolic expression.\n",
+    "\n",
+    "Take for instance the linear equation"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "x = tf.placeholder(dtype=tf.float32, shape=[], name='p')\n",
+    "y = 3 * x + 2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can compute the derivative of `y` with respect to `x` with the `tf.gradients` function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "dy_dx, = tf.gradients(ys=y, xs=[x])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can verify that the gradient evaluates to 3 as expected:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The gradient of y with respect to x is 3.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "with tf.Session() as session:\n",
+    "    dy_dx_val = session.run(dy_dx)\n",
+    "    print('The gradient of y with respect '\n",
+    "          'to x is {}'.format(dy_dx_val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "_Note: Some of you may have noticed that TensorFlow did not complain despite no value being provided for the `x` placeholder. This is because even though `x` is part of the computation graph, the derivative of `y` with respect to `x` does not involve `x`, and therefore evaluating it does not require a value to be passed for `x`._"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Exercise\n",
+    "\n",
+    "Find the minimum of the expression\n",
+    "\n",
+    "$$2(x - 2)^2 + 2(y + 3)^2$$\n",
+    "\n",
+    "using gradient descent by filling in the following code block:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "with tf.Graph().as_default():\n",
```

Comment:
You have not explained `tf.Graph()` before.

I suggest either you explain why you use it, or you write this exercise as a single separate script (with no needs to separate graphs).

Reply:
Good point! Done via 0714eaa.